### DISCUSSION TITLE
Passing JSON objects to and from C#

### DISCUSSION THREAD

**Discussion by AndyHit5687** [cefsharp/CefSharp#3415] [Q&A]
Hi All, I am new to "playing" with CefSharp so am feeling my way. I have successfully embedded it within a WinForm application. I have called a C# method from JavaScript and invoked a delayed callback in JS. I was trying to pass a JSON object into the C# method and added a parameter of type Object to receive it so I could determine the type, it arrives and "GetType()" returns Dictionary'2... I guessed at Dictionary<string,string> but this results in an invalid cast. Can anyone tell me the correct type to use? I would also like to return a value that JS would see as a JSON object.

Also, is there a "promise" equivalent to "IJavascriptCallback"?

public void TestCallback(IJavascriptCallback javascriptCallback, object param)

**Comment by amaitland** [comment]
> I was trying to pass a JSON object into the C# method and added a parameter of type Object

If you passed an object to a method as a parameter then there's no JSON involved. 

The default type for an object is https://docs.microsoft.com/en-us/dotnet/api/system.dynamic.expandoobject?view=net-5.0 Which you can cast to System.Collections.Generic.IDictionary<string,object> should you choose to. You can use the dynamic keyboard with ExpandoObject.

Rather than describe code in future please provide examples of both your JavaScript and c#.

> I would also like to return a value that JS would see as a JSON object.

You can return a POCO and you will get an object in JavaScript. Again no JSON is actually used.

> Also, is there a "promise" equivalent to "IJavascriptCallback"?

No. What exactly are you trying to achieve?

**Reply by amaitland** [reply]
https://github.com/cefsharp/CefSharp/blob/cefsharp/87/CefSharp.Example/JavascriptBinding/AsyncBoundObject.cs#L77
https://github.com/cefsharp/CefSharp/blob/cefsharp/87/CefSharp.Example/Resources/BindingTestAsync.js#L76

**Reply by AndyHit5687** [reply]
Alex,

**Firstly thank you for your prompt helpful reply.** Re "promise" here is some context around what I am trying to do.
I am recently retired from a software engineering role and now have time to try out all of those projects that no one would pay me to do in an effort to keep my mind active. Currently I propose to "just give the results away" if I succeed in making something useful, if I don't then I will have learnt something anyway. I have very little exposure to Javascript, browser programming and C# coming from a C/C++ background so am learning on the fly. I do remember that a javascript library that my previous company used involved the use of "promises" to handle asynchronous callbacks and I noticed that the new object binding calls for CefSharp also use it. I can be a bit of a pedant and I like to at least start off using "best practice" where I can and have assumed that other Javascript programmers will be used to and maybe prefer a "promise" based interface for callbacks but to be honest this is a guess upon my part. I have seen that the addition of a standardized "promise" in JS was welcomed so I was hoping to be able to call a C# method that returned a "promise" which then could be resolved? I am working on a simple application that will, if I ever finish it, present the results of SQL queries using HTML as a front end, in effect a form of report system enhanced by a data manipulation chain something like a very cutdown "SnapLogic" paradigm. I like to use asynchronous data access mechanisms so that the UI remains responsive. My initial thoughts were based upon using a simple asynchronous callback aka "IJavascriptCallback" mechanism but as I said above having seen prior use of promises and that the new CefSharp initial object binding uses promises in async mode I have biased my thoughts to that. I would welcome your opinion as I assume, since you are a maintainer of CefSharp, you are most certainly better informed than me.

Thanks Andy H

**Reply by amaitland** [reply]
>  I was hoping to be able to call a C# method that returned a "promise" which then could be resolved? 

Can you provide a detailed example, pseudo code is fine. More details would be helpful. 

CefSharp is just one of many chromium embedded framework wrappers, as it stands CEF doesn't have first class support for promises, so whilst we use promises where possible,  we have to get creative to actually add support. 

For evaluating JavaScript that returns a promise you can use http://cefsharp.github.io/api/87.1.x/html/Overload_CefSharp_WebBrowserExtensions_EvaluateScriptAsPromiseAsync.htm

**Reply by AndyHit5687** [reply]
Alex,
My plan is, unless it proves impossible, to use the CefSharp integration as follows:
1. Application binds a single C# class, "ReportController".
2. User selects report definition file from file explorer like tree but with reports in it as well as folders.
2. App loads the report definition including the HTML to render the results. The report SQL query and other data manipulations are loaded behind the ReportController.
3. If report has parameters defined the HTML window displays the necessary for the user to fill in any desired values.
4. HTML displays a button or other control to start the report action.
5. When button clicked JavaScript calls method on report controller in async manner thus leaving the UI active. This is where I thought a "promise" would be appropriate. If I can't do this then maybe passing two JS functions, one for success and the other for failure would work as well albeit not "best practice"?
6. On "success" JS calls report controller one or more times to retrieve a "document" object corresponding to what data needs to be displayed. This method can be synchronous as the data should already be ready for consumption.

When it comes to promise declarations I admit to being confused. I have found a reference that shows a promise being declared in JS and then all later examples seem to show that the promise is returned by the asynchronous method, in my desired case the LoadData() C# method. Should I just give up and pass two JS functions as parameters to my LoadData() method or is my understanding all wrong?

**Reply by amaitland** [reply]
> Should I just give up and pass two JS functions as parameters to my LoadData() method or is my understanding all wrong?

I don't have a clear picture of what your code would look like. A code example is worth 1000 words.

I'd suggest getting something working,  then come back with an example and I can make a suggestion on how to improve it.

**Reply by AndyHit5687** [reply]
I will do that. I need a little time to get to the point where I have a working example as I need to learn and resolve things as I go. I will concentrate on the data transfer link between JS and C# and just stub out the actual background data retrieval etc.
Is there a better place to put the example code rather than paste or attach it here or is this an appropriate place?

Thanks Andy H

**Reply by AndyHit5687** [reply]
Alex, just discovered that I can call a JS function with a nested Dictionary<string, object> but cannot return the same. Returning a dictionary raises an exception in "TypeUtils.cpp". Is this by design or omission? If it is the latter would it help if I pulled the latest version of CefSharp and tried to make any necessary changes myself and then sent them back to you for review and hopefully inclusion in the source code? If the latter are there any instructions on how to set up the VS2019 project, that is do I need to download anything extra such as Chrome API stuff in order to compile and link or is CefSharp self contained?

The method below successfully calls my callback and the Chrome debugger sees a JS object like I would expect, even the nested dictionary but changing the return from null to include the same format dictionary raises an exception saying unable to convert. I think this is the line:

//TODO: What exception type?
throw gcnew Exception("Cannot convert object from Cef to CLR.");


        public Dictionary<string, object> TestCallback(IJavascriptCallback javascriptCallback, object param)
        {
            Dictionary<string, object> result = new Dictionary<string, object>
            {
                { "One", "Fluffy" },
                { "Two", 1 } //,
                //{ "Three", new Dictionary<string, object>
                //    {
                //        {"A", 1},
                //        {"B", 2},
                //    }
                //}
            };


            const int taskDelay = 1500;
            IDictionary<string, object> paramObject = (IDictionary<string, object>)param;

            Task.Run(async () =>
            {
                await Task.Delay(taskDelay);

                using (javascriptCallback)
                {
                    Dictionary<string, object> result2 = new Dictionary<string, object>
                    {
                        { "X", "John" },
                        { "Y", 42 },
                        { "Z", new Dictionary<string, object>
                            {
                                {"A", 1},
                                {"B", 2}
                            }
                        }
                    };

                    await javascriptCallback.ExecuteAsync("This callback from C# was delayed " + taskDelay + "ms", result2); //<<== succeeds and JS object created.
                }
            });
            return null; // result; // <<= raises exception in TypeUtils.cpp saying no convertion.
        }

**Reply by amaitland** [reply]
The Async JavaScript Binding implementation should support nested dictionaries. 

https://github.com/cefsharp/CefSharp/blob/cefsharp/87/CefSharp.Example/JavascriptBinding/AsyncBoundObject.cs#L179

The older legacy synchronous JavaScript Binding is not something I have plans to add new features to as per https://github.com/cefsharp/CefSharp/wiki/General-Usage#sync-javascript-binding-jsb

If the change is minimal and includes appropriate qunit tests then I'd consider a PR. 

If you use a POCO then nested properties should already be supported. 

> CefSharp self contained?

Nuget will download the required dependencies. The project uses the Chromium embedded framework, see https://github.com/cefsharp/CefSharp/wiki/General-Usage for background. Chrome doesn't expose an API.

If you need to use the debugger see https://www.chromium.org/developers/how-tos/debugging-on-windows specially the Manually attaching to a child process section. 

Pick the x86 platform from the menu in VS.

**Reply by AndyHit5687** [reply]
Alex,

I modified my C# object registration to:

            m_chromeBrowser.JavascriptObjectRepository.Settings.LegacyBindingEnabled = true;
            m_chromeBrowser.JavascriptObjectRepository.Register("ReportController", new ReportController(m_chromeBrowser, this), isAsync: true, options: BindingOptions.DefaultBinder);

And re-instated the return of a nested dictionary object. No exception now so good. Using the chrome debugger I discovered that JS call to my C# function returned a "promise" so I changed the JS code to this just to try it out.

        function InstigateCallback() {
            var returnResult;

//            returnResult = ReportController.testCallback(MyCallback, person);
            returnResult = ReportController.testCallback(MyCallback, person);
            returnResult.then(function (returnObject) {
                alert("THEN");
            }).catch(function () {
                    alert("CATCH");
            });
        }
        function MyCallback(msg, result) {
            alert("My callback returned - " + msg + " X = " + result.X + " Y = " + result.Y);
        }
where the "then" function parameter "returnObject" is a JS object that represents the dictionary returned from C#.

I don't know where to look but I assume you have built in some promise support. I then added a Thread.Sleep(10000) into my C# code to stub a lengthy operation before returning and placed another alert after the promise code:

        function InstigateCallback() {
            var returnResult;

//            returnResult = ReportController.testCallback(MyCallback, person);
            returnResult = ReportController.testCallback(MyCallback, person);
            returnResult.then(function (returnObject) {
                alert("THEN");
            }).catch(function () {
                    alert("CATCH");
            });
            alert("XXX");
        }
        function MyCallback(msg, result) {
            alert("My callback returned - " + msg + " X = " + result.X + " Y = " + result.Y);
        }

And the alerts occurred "XX" -> "My callback..." -> "THEN" which is what I wanted. I haven't tried making another C# call before the promise is fulfilled.

**Reply by amaitland** [reply]
> Also, is there a "promise" equivalent to "IJavascriptCallback"?

In version `88.2.90` I've added manual support for resolving a promise when calling `IJavascriptCallback.ExecuteAsync`. It's a manual process and requires some extra handling, I've added an example to 
https://github.com/cefsharp/CefSharp/wiki/Advanced-Async-JavaScript-Binding-(JSB)#javascript-callback-with-promise
